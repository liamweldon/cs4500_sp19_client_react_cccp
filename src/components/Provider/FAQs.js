import React from 'react';
import './Provider.scss';

const FAQs = ({faqs}) => (
  <div className="faqs">
    {faqs.map((faq) => (
      <div className="faq">
        <h5>{faq.question}</h5>
        <div>{faq.answer}</div>
      </div>
    ))}
  </div>
);

export default FAQs;
