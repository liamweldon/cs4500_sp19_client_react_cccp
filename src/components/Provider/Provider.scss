.provider {
  margin-left: 7em;
  margin-right: 7em;
  margin-top: 3em;
  margin-bottom: 1em;

  .section {
    border-bottom: 1px solid lightgray;
    padding-bottom: 1em;
    padding-top: 1em;
  }

  img {
    margin-right: 1em;
    margin-bottom: 1em;
  }

  i {
    margin-right: 0.5em;
  }

  .row {
    display: flex;
    margin: 0; // override default margin -15
    margin-bottom: 1em;
  }

  .col {
    display: flex;
    flex-direction: column;
  }

  .faqs {
    .faq {
      border-bottom: 1px solid lightgray;
      padding-bottom: 1em;
      padding-top: 1em;
    }
  }

  .reviews {
  }

  .yellow {
    color: #ffe000;
  }
}
