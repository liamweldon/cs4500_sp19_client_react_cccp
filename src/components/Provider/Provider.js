import React from 'react';
import './Provider.scss';
import FAQs from './FAQs';

const Provider = ({provider, faas}) => (
  <div className="provider">
    <div className="row">
      <ul className="nav nav-pills">
        <li className="nav-item">
          <a className="nav-link" href="#about">
            About
          </a>
        </li>
        <li className="nav-item">
          <a className="nav-link" href="#reviews">
            Reviews
          </a>
        </li>
        <li className="nav-item">
          <a className="nav-link" href="#faqs">
            FAQs
          </a>
        </li>
      </ul>
    </div>
    <div className="row">
      <img src="https://picsum.photos/150/150" alt="car" />
      <h2 className="name">{provider.username}</h2>
    </div>
    <div className="description row">{provider.description}</div>
    <div className="row section" id="about">
      <div className="col overview">
        <h4> Overview </h4>
        <label>
          <i className="fa fa-trophy" />
          Hired {provider.timesHired} times
        </label>
        <label>
          <i className="fa fa-lock" />
          {provider.isBackgroundChecked ? 'Background checked' : ''}
        </label>
        <label>
          <i className="fa fa-users" />
          {provider.numEmployees} Employees
        </label>
        <label>
          <i className="fa fa-briefcase" />
          {provider.numYearsInBusiness} Years in business
        </label>
      </div>
      <div className="col payment">
        <h4> Payment Methods </h4>
      </div>
    </div>
    <div className="section" id="reviews">
      <h4> Reviews </h4>
    </div>
    <div className="row" id="faqs" />
    <h4> FAQs </h4>
    <FAQs faqs={faas} />
  </div>
);

export default Provider;
