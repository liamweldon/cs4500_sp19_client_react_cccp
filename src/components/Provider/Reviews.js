import React from "react";
import _ from "lodash";
import faker from "faker";
import moment from "moment";
import "./Provider.scss";
import Rating from "./Rating";

const reviewSortOptions = [
  "Recommended",
  "Highest Rated",
  "Lowest Rated",
  "Newest Review",
  "Oldest Review"
];

const reviewsPerPage = 2;

class Reviews extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      reviews: [],
      sortedBy: reviewSortOptions[0],
      currPage: 0,
      provider: {
        ratingScores: [],
      }
    };
  }

  componentDidMount() {
    this.initializeMockReviews();
  }

  initializeMockReviews() {
    const mockReviews = [];
    const pushRandomReview = () => {
      const review = {};
      // review:
      // User name
      // rating: int 1-5
      // Date: date
      // text: String
      // response: STring
      review.id = faker.random.number();
      review.userName = faker.name.findName();
      review.score = faker.random.number({ max: 5, min: 1 });
      review.date = faker.date.recent(20);
      review.reviewText = faker.lorem.sentences();
      review.providerReply = faker.lorem.sentence();
      mockReviews.push(review);
      const provider = {};

      provider.ratingScores = [93, 24, 12, 10, 4];

      this.setState({ reviews: mockReviews, provider });
    };

    _.times(10, pushRandomReview);
  }

  getSortedReviews() {
    return this.state.reviews;
  }

  renderHeader() {
    const { provider } = this.state;
    return (
      <div>
        <div className="row">
          <div className="col-8">
            <a name="reviews" />
            <h3>Reviews</h3>
          </div>
          <hr />
          <div className="col-4">
            <select className="form-control float-right">
              <option>Recommended</option>
              <option>Highest rated</option>
              <option>Lowest rated</option>
              <option>Newest first</option>
              <option>Oldest first</option>
            </select>
            <br />
            <br />
          </div>
        </div>
        <div className="row">
          <div className="col-4">
            <h4>4.4</h4>
            {this.renderStars(4)}
            <br />
            {this.state.reviews.length} reviews
          </div>
          <div className="col-8">
            {provider.ratingScores.map((score, index) => (
              <Rating key={index} index={5 - index} score={score} />
            ))}
          </div>
        </div>
        <br />
      </div>
    );
  }

  renderStars(n) {
    return _.times(n, i => {
      return <i className={"fa fa-star yellow"} key={i} />;
    });
  }

  renderReviews() {
    const { currPage } = this.state;
    const firstIndex = currPage * 2;
    return (
      <div>
        {_.map(
          this.getSortedReviews().slice(firstIndex, firstIndex + 2),
          review => {
            return (
              <div key={review.id}>
                <div>
                  <h6>{review.userName}</h6>
                </div>
                <br />
                <div>{this.renderStars(review.score)}</div>
                <br />
                <div>{review.reviewText}</div>
                <br />
                <div>{moment(review.data).format("MMM D, YYYY")}</div>
                <br />
                <div className="alert alert-secondary" role="alert">
                  {review.providerReply}
                </div>
                <hr />
              </div>
            );
          }
        )}
      </div>
    );
  }

  renderPaginator() {
    const { reviews, currPage } = this.state;

    const pages = Math.round(reviews.length / 2);

    return (
      <div className="center">
        <nav aria-label="...">
          <ul className="pagination">
            <li className={`page-item ${currPage === 0 ? "disabled" : null}`}>
              <span className="page-link">Previous</span>
            </li>
            {_.times(pages, i => {
              return (
                <li
                  key={i}
                  className={`page-item ${currPage === i ? "active" : null}`}
                  aria-current="page"
                  onClick={() => this.setState({ currPage: i })}
                >
                  <span className="page-link">
                    {i + 1}
                    <span className="sr-only">(current)</span>
                  </span>
                </li>
              );
            })}
            <li
              className={`page-item ${
                currPage === pages - 1 ? "disabled" : null
              }`}
              onClick={() => this.setState({ currPage: currPage + 1 })}
            >
              <a className="page-link">Next</a>
            </li>
          </ul>
        </nav>
      </div>
    );
  }

  render() {
    return (
      <div id="reviews">
        {this.renderHeader()}
        {this.renderReviews()}
        {this.renderPaginator()}
      </div>
    );
  }
}

export default Reviews;
