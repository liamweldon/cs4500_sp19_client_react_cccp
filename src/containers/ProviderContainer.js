import React from 'react';
import UserService from '../services/UserService';
import FAQAnswerService from '../services/FAQAnswerService';
import '../components/table.scss';
import '../components/FAQs.scss';
import Provider from '../components/Provider/Provider.js';

class ProviderContainer extends React.Component {
  constructor(props) {
    super(props);
    this.userService = UserService.getInstance();
    this.faaService = FAQAnswerService.getInstance();
    this.state = {provider: {}, faas: []};
  }

  getProvider = () => {
    // TODO in the future, we will find by the ID in the path
    this.userService.findUserById('522').then((provider) => {
      if (provider) {
        this.setState({
          provider
        });
      }
    });
  };

  getFAAs = () => {
    // in the future, we will find by the provider's ID in the path
    this.faaService.findFAAsByUser('522').then((faas) => {
      if (faas) {
        this.setState({
          faas
        });
      }
    });
  };

  componentDidMount() {
    this.getProvider();
    this.getFAAs();
  }

  render() {
    return <Provider provider={this.state.provider} faas={this.state.faas} />;
  }
}

export default ProviderContainer;
