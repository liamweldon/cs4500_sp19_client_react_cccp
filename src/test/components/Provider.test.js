import React from 'react';
import Provider from '../../components/Provider/Provider';
import ProviderContainer from '../../containers/ProviderContainer';
import renderer from 'react-test-renderer';
import faasJson from '../mockData/faa.mock.json';
import providerJson from '../mockData/provider.mock.json';
import {StaticRouter} from 'react-router';
import Enzyme, {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({adapter: new Adapter()});

test('Provider Summary Snapshots', () => {
  const component = renderer.create(
    <StaticRouter location="/provider" context={{}}>
      <Provider faas={faasJson} provider={providerJson} />
    </StaticRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Provider summary DOM', () => {
  let wrapper = mount(
    <StaticRouter location="/provider" context={{}}>
      <ProviderContainer />
    </StaticRouter>
  );

  let providerComponent = wrapper.find(ProviderContainer);

  providerComponent.getProvider = jest.fn().mockReturnValue(providerJson);
  providerComponent.getFAAs = jest.fn().mockReturnValue(faasJson);
  wrapper.update();
  providerComponent.setState({faas: faasJson});
  providerComponent.setState({provider: providerJson});

  expect(wrapper.find('.provider .faq').length).toEqual(faasJson.length);
  expect(wrapper.find('.provider .description').text()).toEqual(providerJson.description);
  expect(wrapper.find('.provider .name').text()).toEqual(providerJson.firstName);
});
