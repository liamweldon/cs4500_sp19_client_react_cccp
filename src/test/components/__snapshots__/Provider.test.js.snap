// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider Summary Snapshots 1`] = `
<div
  className="provider"
>
  <div
    className="row"
  >
    <ul
      className="nav nav-pills"
    >
      <li
        className="nav-item"
      >
        <a
          className="nav-link"
          href="#about"
        >
          About
        </a>
      </li>
      <li
        className="nav-item"
      >
        <a
          className="nav-link"
          href="#reviews"
        >
          Reviews
        </a>
      </li>
      <li
        className="nav-item"
      >
        <a
          className="nav-link"
          href="#faqs"
        >
          FAQs
        </a>
      </li>
    </ul>
  </div>
  <div
    className="row"
  >
    <img
      alt="car"
      src="https://picsum.photos/150/150"
    />
    <h2
      className="name"
    >
      demi
    </h2>
  </div>
  <div
    className="description row"
  >
    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo.
  </div>
  <div
    className="row section"
    id="about"
  >
    <div
      className="col overview"
    >
      <h4>
         Overview 
      </h4>
      <label>
        <i
          className="fa fa-trophy"
        />
        Hired 
        3
         times
      </label>
      <label>
        <i
          className="fa fa-lock"
        />
        Background checked
      </label>
      <label>
        <i
          className="fa fa-users"
        />
        9
         Employees
      </label>
      <label>
        <i
          className="fa fa-briefcase"
        />
        2
         Years in business
      </label>
    </div>
    <div
      className="col payment"
    >
      <h4>
         Payment Methods 
      </h4>
    </div>
  </div>
  <div
    className="section"
    id="reviews"
  >
    <h4>
       Reviews 
    </h4>
  </div>
  <div
    className="row"
    id="faqs"
  />
  <h4>
     FAQs 
  </h4>
  <div
    className="faqs"
  >
    <div
      className="faq"
    >
      <h5>
        Do you offer discounted services?
      </h5>
      <div>
        No
      </div>
    </div>
    <div
      className="faq"
    >
      <h5>
        How many employees does your company have?
      </h5>
      <div>
        80
      </div>
    </div>
  </div>
</div>
`;
